### ISSUE TITLE
bapbuild failing?

### ISSUE THREAD

**Issue by Dinatrix** [BinaryAnalysisPlatform/bap-plugins#7]
Hi!
I was able to use bapbuild and bap on the hello_world example, but when I attempted it on a sample plugin, I'm not able to get past the bapbuild stage?

I have here the checkpath.ml file from your bap-plugins saved and try to run it with 'bapbuild checkpath.plugin' on the command line.

![checkpath](https://cloud.githubusercontent.com/assets/10760263/12646918/c7c6ec00-c5a0-11e5-908a-937d21af4c7a.PNG)

![bapbuild](https://cloud.githubusercontent.com/assets/10760263/12646911/c21af18e-c5a0-11e5-864b-15e06a5fc82c.PNG)

So many errors fill the screen that the original command can't be seen when I scroll up. Most have to do with files having a .cma/.cmx/.cmxa/.cmi suffix? It's curious the hello_world example would work but not this. Would appreciate any insight into what may be causing this. Thanks!

**Comment by ivg** [comment]
TL;DR; Don't run `bapbuild` or `ocamlbuild` in a home folder. Create a directory with your project and run it there. 

`ocamlbuild` (and thus `bapbuild`) is very picky to the hygiene. I will tell about `ocamlbuild`, as the same is true for `bapbuild`. 

`ocamlbuild` works in a folder, it scans all files recursively deep, and tries to compile it. In order to prevent building something wrong or interacting with other build systems, ocamlbuild imposes a hygiene requirement, that there is no binary objects files ( `.o`, `.cm*` etc). Since you're building in you home folder, ocamlbuild assumes, that your whole home hierarchy is some sort of big project, that you would like to build. He scans the systems and finds lots of objects files and libraries in `.opam` folder, that holds all ocaml universe. 

Note: I really hope, that you didn't follow the advice to call `_build/hygiene.sh`, as this will ruin you `.opam` folder, and may break other projects in your home folder. If you did this, then it is better to reinstall opam.

**Comment by Dinatrix** [comment]
After creating a separate folder for the project and adding an empty _tags file, there seems to be an issue with checkpath.ml itself?  Which is strange considering it is just a copy/paste of what I found in bap-plugins.

![checkpathml](https://cloud.githubusercontent.com/assets/10760263/12648715/47861f44-c5a9-11e5-9b2c-6a6d97f77d50.PNG)

And the line in question:

![image](https://cloud.githubusercontent.com/assets/10760263/12648741/68197f1c-c5a9-11e5-9dac-0287144e6519.png)



Thanks again!

**Comment by ivg** [comment]
now its ok :) The `checkpath` plugin is broken, as it wasn't updated  to the latest version by the author, an he will be punished to the full extent of the law. (disclaimer I'm the author). 

There is no more `Block.dfs` function, as it was superseded, by a `Block.to_graph` function, that returns a graph representation of a CFG, with much richer interface. 

Whatever was called `dfs` now, can be implemented using new interface as follows:

```ocaml
let dfs blk =
  Block.to_graph blk |>
  Graphlib.postorder_traverse (module Block.Graph)
```

with this function, you need just substitute all `Block.dfs` with this `dfs` function, e.g., 
```ocaml
let goto src point : block option =
  dfs src |> Seq.find ~f:(fun blk -> Addr.(point = Block.addr blk))
```

P.S. I'm not fixing it right now, as I'm preparing a new release, that will require few more changes to this plugin. So I will update all plugins at once after the release.

**Comment by ivg** [comment]
P.S. This plugin is rather naive, we have a much more interesting plugin that can validate paths and much more. Currently it is a work in progress, that is stored in `quarantine` folder, and it currently compiles only against specific branch of bap. But I will (I actually must) release it before Monday, so it will watch for updates.

**Comment by ivg** [comment]
btw, you don't need to create `_tags` file. The message, from `ocamlbuild` is sort of heuristics, that fails in our case, so it is misleading. I pruned this message, so the new `bapbuild` will not mislead you.

**Comment by Dinatrix** [comment]
Great! Will check back Monday :D

**Comment by Dinatrix** [comment]
Hi! Just checking back in to see if the updated plugin you mentioned in quarantine was released? The quarantine folder is currently showing 3 months as the most recent commit but maybe it was uploaded elsewhere?
Thanks!

**Comment by ivg** [comment]
Hi! Unfortunately I was refocused on a different task, and the work is still unreleased (not merged into the master branch). If you're really interested, then I can work with you tomorrow or later, to obtain a working version of quarantine. For a current master branch, or, at least provide commit references where it did work.

**Comment by Dinatrix** [comment]
That would be great! I don't think I mentioned, but I'm working to create a demo of BAP that Apogee can present so am eager to learn more about quarantine. Would you recommend I contact you via email to work that out or via github?

**Comment by ivg** [comment]
Sure, you can use any method that suits you better, email, gitter or anything else. I've updated quarantine so that it can work for the latest master version. In nowadays you can compile the latest master version with `opam pin add bap --dev-repo’, and then go and pull latest quarantine and just do `make clean; make test` in its folder. Ideally everything should be ok (literally). If it doesn't then make sure that you've IDA installed.

**Comment by ivg** [comment]
To conclude. The quarantine plugin doesn't no longer exist. It gave birth to a new library [microx](https://github.com/BinaryAnalysisPlatform/bap/pull/377) and to a [taint propagation framework](https://github.com/BinaryAnalysisPlatform/bap/pull/379), and [saluki framework](https://github.com/BinaryAnalysisPlatform/bap-plugins/tree/master/saluki). The first two are now part of bap.  To get an alpha version of bap follow this instructions:

1. You need to add our repository, this should be done only once:

     ```
     opam repository add bap git://github.com/BinaryAnalysisPlatform/bap.git
    ```

2. Don't forget to update the repository. We push new packages very often.

    ```
    opam update
    ```

3. Now install bap:

    ```
    opam install bap
   ```

   This is a meta-package that will install essentials part of BAP.

4. Saluki is not yet a part of BAP distribution. To get it just do the following:

    ```
    opam install bap-saluki
    ```
